### PR TITLE
Added option to center image when loading particle emission mask

### LINKED ISSUE #16411: 2D Particle Emission Mask is offset
**Godot version:**

3.0
**OS/device including version:**

**Issue description:**

When I load an Emission Mask for a 2DParticle system the result is offset since the mask is placed from the top left corner and not the center.
In the screenshots below the Emission Mask has been generated from the same image used for the Well sprite.
The offset might be the intended behavior or a limitation of the way the particle system works now I don't know. Placement from the center would seem to be the expected behavior however.

The first image shows the center of the Well sprite.
The second image shows the center of the 2DParticle system. 

godotparticlemaskbug2

godotparticlemaskbug1

**Steps to reproduce:**
Place a 2D Sprite and a 2DParticle system at 0,0. Add an image to the sprite and load the same image as an Emission Mask in the 2DParticle system.

**Minimal reproduction project:**

### PR DESCRIPTION
**Pull request by bgie**
When loading an image to use as a particle emission mask, the option to center the image is now available. The centering is applied when generating the mask, so no extra runtime cost is involved.

Fixes https://github.com/godotengine/godot/issues/16411

### CODE REVIEW

**Comment by bgie** [comment]
First contribution to the project, so do let me know if I missed some guideline, would be happy to fix it.

File: editor/plugins/gpu_particles_2d_editor_plugin.cpp [hunk 1]
```diff
@@ -292,11 +292,16 @@ void GPUParticles2DEditorPlugin::_generate_emission_mask() {
 	texdata.resize(w * h * 2 * sizeof(float));
 
 	{
+		Vector2 offset(0., 0.);
```
Comment (Calinou) [inline editor/plugins/gpu_particles_2d_editor_plugin.cpp:295]: The default value is already `Vector2(0, 0)`, so you can skip the assignment. Also, we generally don't use C++-style parentheses initializers and prefer to use `=` instead where we do need assignment.

```suggestion
		Vector2 offset;
```

File: editor/plugins/cpu_particles_2d_editor_plugin.cpp [hunk 2]
```diff
@@ -217,11 +217,16 @@ void CPUParticles2DEditorPlugin::_generate_emission_mask() {
 	}
 
 	{
+		Vector2 offset(0., 0.);
```
Comment (Calinou) [inline editor/plugins/cpu_particles_2d_editor_plugin.cpp:220]: The default value is already `Vector2(0, 0)`, so you can skip the assignment. Also, we generally don't use C++-style parentheses initializers and prefer to use `=` instead where we do need assignment.

```suggestion
		Vector2 offset;
```

File: editor/plugins/gpu_particles_2d_editor_plugin.cpp (hunk 1 above)
Comment (bgie) [reply editor/plugins/gpu_particles_2d_editor_plugin.cpp]: I fixed it with an amend... and then noticed that github has an easy button to commit your suggestion directly...

File: editor/plugins/cpu_particles_2d_editor_plugin.cpp (hunk 2 above)
Comment (bgie) [reply editor/plugins/cpu_particles_2d_editor_plugin.cpp]: Same... thank you.

**Comment by KoBeWi** [comment]
Having a separate section for each checkbox is a bit awkward.
image
I think Centered and Capture from Pixel should be under one section, although the latter might need to be renamed. Something like

**Options**
- Centered
- Capture Colors from Pixel

**Comment by Calinou** [comment]
@bgie It looks like you force-pushed your branch in a way that makes it have no commits, which means GitHub automatically closed the pull request.

You need to open a new pull request with a different branch name – do not use your fork's `master` branch as your branch for pull requests.

**Comment by bgie** [comment]
@Calinou Thanks for the clarification and clear instructions. 
I had noticed I was working on master and tried to correct that, not considering that the pull request is apparently (obviously) linked to a branch name.

**Pull request closed without merging**